src/libraries/System.Linq/src/System/Linq/Last.cs

```diff
@@ -117,7 +121,7 @@ public static TSource LastOrDefault<TSource>(this IEnumerable<TSource> source, F
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.predicate);
             }
 
-            if (source is OrderedEnumerable<TSource> ordered)
+            if (source is OrderedIterator<TSource> ordered)
```

Comment:
Out of curiosity, why specialize only for `OrderedIterator` and not `Iterator`? Assuming that every implementation provides a bespoke `TryGetLast` implementation shouldn't we always be delegating to that?

Reply:
This is a TryGetLast that's specific to OrderedIterator, taking a `Func<TSource, bool>` predicate. If we ever push that down to Iterator, then we could certainly extend this to use the base as well.